lib/tennis_score.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# abstract class used to manage basic elements of scoring in
+# all stages of a tennis match
+class TennisScore
+  attr_accessor :points, :winner, :players
+
+  def initialize(min_win, win_by, type, players)
+    # create score array with scores for both teams
+    @points = [0, 0]
+    @min_win = min_win # defines the minimum points a player must have to win
+    @win_by = win_by # defines the number of points more a player must have than their opposition to win
+    @type = type # usually game/set/match
+    @players = players # the two player objects for the match
+  end
+
+  # converts the score into a the tennis scores
+  def score
+    "#{@type.capitalize} score:\n#{@points[0]} : #{@points[1]}"
+  end
+
+  def win?
+    true unless which_player_win.nil?
+  end
+
+  def which_player_win
```

Comment:
Alarm bells ring when I feel the need to write a truth table to work out what a boolean condition is doing.
If we try to make this more readable I would be initially inclined to refactor it into
```
def which_player_win
return if (@points[0] - @points[1]).abs < @win_by
return if (@points[0] < @min_win && @points[1] < @min_win)
@winner = @points.index(@points.max)
end
```
I would then maybe extract the code into methods so that it is more readable

```
def which_player_win
return if scores_level
return if no_winning_score
@winner = @points.index(@points.max)
end

private
def scores_level
(@points[0] - @points[1]).abs < @win_by
end

def winning_score
(@points[0] < @min_win && @points[1] < @min_win)
end
```

The advantage of this is that when a new developer looks at the code they can easily see what the which_player_wins method is doing without having to understand the complexity of the private methods.